File: 2786/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(maxScore([]int{2,3,6,1,9,2}, 5))
}

func maxScore(nums []int, x int) int64 {
	res := int64(nums[0])
	dp := [2]int64{math.MinInt32, math.MinInt32}
	dp[nums[0] % 2] = int64(nums[0])
	for i := 1; i < len(nums); i++ {
		par := nums[i] % 2
		cur := max(dp[par] + int64(nums[i]), dp[1 - par] - int64(x) + int64(nums[i]))
		res = max(res, cur)
		dp[par] = max(dp[par], cur)
	}
	return res
}

File: 2786/main.rs
struct Solution{}

impl Solution {
    pub fn max_score(nums: Vec<i32>, x: i32) -> i64 {
        let mut res = nums[0] as i64;
        let mut dp: Vec<i64> = vec![i32::MIN as i64, i32::MIN as i64];
        dp[(nums[0] % 2) as usize] = nums[0] as i64;
        for i in 1..nums.len() {
            let par = (nums[i] % 2) as usize;
            let cur = (dp[par] + nums[i] as i64).max(dp[1-par] - x as i64 + nums[i] as i64);
            res = res.max(cur);
            dp[par] = dp[par].max(cur);
        }
        return res;
    }
}

fn main() {
    println!("{}", Solution::max_score(vec![2,3,6,1,9,2], 5))
}